### PR TITLE
Add basic implementation of Environment.FailFast

### LINKED ISSUE #14827: [netcore]Implement Environment.FailFast
See the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.environment.failfast?view=netcore-3.0) for full description.

Couple of call outs:
1.) The process should write out a crash dump before terminating.  It would be nice if we could use our crash reporter as I think it does that plus a bunch of other useful things.
2.) On windows, we should write to the event log.  I don't know what the behavior is on other platforms.

### PR DESCRIPTION
**Pull request by filipnavara**
Fixes #14827.

### CODE REVIEW

File: mono/metadata/icall.c [hunk 1]
```diff
@@ -7391,6 +7391,28 @@ ves_icall_System_Environment_Exit (int result)
...
 
+void
+ves_icall_System_Environment_FailFast (MonoStringHandle message, MonoExceptionHandle exception, MonoStringHandle errorSource, MonoError *error)
+{
+	if (MONO_HANDLE_IS_NULL (message)) {
+		g_warning ("CLR: Managed code called FailFast without specifying a reason.");
+	} else {
+		char *msg = mono_string_handle_to_utf8 (message, error);
+		g_warning ("CLR: Managed code called FailFast, saying \"%s\"", msg);
+		g_free (msg);
```
Comment (jaykrell) [inline mono/metadata/icall.c]: On Windows please consider `%ls` and passing through the Unicode.
i.e. no alloc, copy, free.
For that matter, if `sizeof(wchar_t) == sizef(gunichar2)`, maybe something.

Comment (jaykrell) [reply mono/metadata/icall.c]: Also please consider a small, maybe 64 local buffer, init with 0 and copy the first 64 or length chars from message.
i.e. if a dump is made, and includes stack, but doesn't follow pointers -- I don't know.

File: netcore/System.Private.CoreLib/src/System/Environment.cs [hunk 2]
```diff
@@ -54,18 +54,16 @@ partial class Environment
...
-			throw new NotImplementedException ();
+			FailFast (message, exception, null);
 		}
 
-		public static void FailFast (string message, Exception exception, string errorMessage)
-		{
-			throw new NotImplementedException ();
-		}
+		[MethodImplAttribute (MethodImplOptions.InternalCall)]
+		public extern static void FailFast (string message, Exception exception, string errorSource);
```
Comment (jaykrell) [inline netcore/System.Private.CoreLib/src/System/Environment.cs]: Does C# have default parameters? Can you say just one function:
```
public extern static void FailFast (string message, Exception exception = null, string errorSource = null);
```

File: mono/metadata/icall.c (hunk 1 above)
Comment (filipnavara) [reply mono/metadata/icall.c]: Agreed on both points. I just wanted to have something working first, pass the CI and then improve upon that.

File: netcore/System.Private.CoreLib/src/System/Environment.cs (hunk 2 above)
Comment (filipnavara) [reply netcore/System.Private.CoreLib/src/System/Environment.cs]: This is public API, the shape cannot be changed.

File: mono/metadata/icall.c [hunk 3]
```diff
@@ -7391,6 +7391,28 @@ ves_icall_System_Environment_Exit (int result)
...
+
+	if (!MONO_HANDLE_IS_NULL (exception)) {
+		mono_print_unhandled_exception_internal ((MonoObject *) MONO_HANDLE_RAW (exception));
+	}
+
+	// NOTE: While this does trigger WER on Windows it doesn't quite provide all the
+	// information in the error dump that CoreCLR would. On Windows 7+ we should call
+	// RaiseFailFastException directly instead of relying on the C runtime doing it
+	// for us and pass it as much information as possible.
+	abort ();
```
Comment (jaykrell) [inline mono/metadata/icall.c]: Windows 7 is the minimum requirement for a while now. Feel free to depend on it.
Notice that importing code from CoreFX was making it so.
And Vista is reall quite old now.

File: mono/metadata/icall.c (hunk 1 above)
Comment (jaykrell) [reply mono/metadata/icall.c]: Looks good despite all my nitpicking.

File: mono/metadata/icall.c (hunk 3 above)
Comment (jaykrell) [reply mono/metadata/icall.c]: If we are going to mention things, mention Windows 8 __failfast. Maybe another year or two down the line that'll be it.

Comment (filipnavara) [reply mono/metadata/icall.c]: I knew it was minimum requirement for .NET Core but I was unsure about Mono. I wrote the relevant part in managed code but I need to translate it back to native code and test the dumps.

Comment (jaykrell) [reply mono/metadata/icall.c]: Almost a year ago.
https://github.com/mono/mono/pull/9568

File: netcore/CoreFX.issues.rsp [hunk 4]
```diff
@@ -175,16 +175,6 @@
 # https://github.com/mono/mono/blob/18f0bf055eb2bfa8981062bdad3af2fb4bbc9be4/netcore/System.Private.CoreLib/shared/System/String.cs#L811-L821
 -nomethod System.Tests.StringTests.NormalizationTest
 
-# Environment.FailFast is not implemented
```
Comment (EgorBo) [inline netcore/CoreFX.issues.rsp]: These tests are unrelated actually - they call FailFast because of Debug.Assert (in debug mode)
Something wrong with fp operations inside number formatting

Comment (filipnavara) [reply netcore/CoreFX.issues.rsp]: Interesting. I am pretty sure they passed when I removed the exclusions, perhaps for unrelated reasons.

Comment (filipnavara) [reply netcore/CoreFX.issues.rsp]:
```
  Discovering: System.Memory.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Memory.Tests (found 2295 of 2347 test cases)
  Starting:    System.Memory.Tests (parallel test collections = on, max threads = 4)
  Finished:    System.Memory.Tests
=== TEST EXECUTION SUMMARY ===
   System.Memory.Tests  Total: 45383, Errors: 0, Failed: 0, Skipped: 0, Time: 26.037s
```

Comment (EgorBo) [reply netcore/CoreFX.issues.rsp]: Did you run them in debug mode? (`make debug-bcl && MONO_ENV...`)

Comment (filipnavara) [reply netcore/CoreFX.issues.rsp]: Nope... good point.

Comment (filipnavara) [reply netcore/CoreFX.issues.rsp]: I can reproduce it now. I'll put the exclusions back and clarify the comment. Thanks!

File: mono/metadata/icall.c [hunk 5]
```diff
@@ -7391,6 +7391,29 @@ ves_icall_System_Environment_Exit (int result)
...
+
+	if (!MONO_HANDLE_IS_NULL (exception)) {
+		mono_print_unhandled_exception_internal ((MonoObject *) MONO_HANDLE_RAW (exception));
+	}
+
+	// NOTE: While this does trigger WER on Windows it doesn't quite provide all the
+	// information in the error dump that CoreCLR would. On Windows 7+ we should call
+	// RaiseFailFastException directly instead of relying on the C runtime doing it
+	// for us and pass it as much information as possible. On Windows 8+ we can also
+	// use the __failfast intrinsic.
```
Comment (CoffeeFlux) [inline mono/metadata/icall.c]: I believe it's `__fastfail`, because consistent naming is overrated. Also, will `__fastfail` still trigger WER?

Comment (filipnavara) [reply mono/metadata/icall.c]: Damn, will fix. It is documented to trigger WER.

File: mono/metadata/icall.c (hunk 1 above)
Comment (NattyNarwhal) [reply mono/metadata/icall.c]: Shouldn't this use `syslog` to emulate Event Log semantics on WIndows?

Comment (filipnavara) [reply mono/metadata/icall.c]: This is not the part that is supposed to be written to event log. The event log should be written roughly at the `abort ();` call. In fact it happens on Windows that way and you get link to the WER minidump if enabled.

Printing this message is largely supposed to offer some indication either through attached native debugger, or through system log if no debugger is attached. On CoreCLR this goes through `OutputDebugString` on Windows. `syslog` may or may not be a good substitute, not sure about it.

`g_warning` is not perfect but it gets the job done here (both developer and user can see the message) and it has the added bonus of being wired properly on Android builds.

**Comment by lewing** [comment]
@monojenkins build failed
